File: 2/req.go
/*
 * @Author       :  Ayouth
 * @Date         :  2022-06-07 GMT+0800
 * @LastEditTime :  2022-06-07 GMT+0800
 * @FilePath     :  req.go
 * @Description  :	GO Http异步请求测试
 * Copyright (c) 2022 by Ayouth, All Rights Reserved.
 */
package main

import (
	"flag"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"
)

func timer(fn func()) {
	start := time.Now()
	fn()
	dur := time.Since(start)
	fmt.Printf("time spent:\t%.2fs", dur.Seconds())
}

func req(client http.Client, url, method string) {
	req, err := http.NewRequest(method, url, nil)
	if err != nil {
		fmt.Println(err)
		return
	}
	req.Header.Set(
		"User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:101.0) Gecko/20100101 Firefox/101.0",
	)
	req.Header.Set(
		"Accept", "*/*",
	)
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println(err)
	} else if resp.Status != "200 OK" {
		fmt.Println("Status Error:", resp.Status)
	}
}

func run(url, method string, count int, limit int) {
	method = strings.ToUpper(method)
	client := http.Client{}
	wg := sync.WaitGroup{}
	ch := make(chan struct{}, limit)
	for i := 0; i < count; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			ch <- struct{}{}
			req(client, url, method)
			<-ch
		}()
	}
	wg.Wait()
	fmt.Printf("url:\t%s\nmethod:\t%s\ncount:\t%d\nlimit:\t%d\n", url, method, count, limit)
}

func main() {

	var url string
	var method string
	var count int
	var limit int
	flag.StringVar(&url, "u", "http://127.0.0.1:8080/", "目标URL端口，默认http://127.0.0.1:8080/")
	flag.StringVar(&method, "m", "GET", "请求格式，默认GET")
	flag.IntVar(&count, "c", 500, "请求数量，默认500")
	flag.IntVar(&limit, "l", 64, "最大并发数，默认64")
	flag.Parse()

	timer(func() {
		run(
			url,
			method,
			count, // 请求数量
			limit, // 最大并发数
		)
	})
	// 请求本地构建的服务器
	// timer(func() {
	// 	run(
	// 		"http://127.0.0.1:8080/",
	// 		"GET",
	// 		10000, // 请求数量
	// 		64,    // 最大并发数
	// 	)
	// })

	// 请求远程服务器
	// timer(func() {
	// 	run(
	// 		"https://www.jb51.net/",
	// 		"GET",
	// 		500, // 请求数量
	// 		64,  // 最大并发数
	// 	)
	// })

}

File: 2/req.py
# coding=utf-8

"""
@Author       :  Ayouth
@Date         :  2022-06-07 GMT+0800
@LastEditTime :  2022-06-07 GMT+0800
@FilePath     :  req.py
@Description  :  Python 异步http请求测试
@Copyright (c) 2022 by Ayouth, All Rights Reserved. 
"""
import asyncio
import time
import aiohttp
import os


async def make_request(client: aiohttp.ClientSession, url: str, method: str) -> None:
    async with client.request(url=url, method=method) as resp:
        if resp.status != 200:
            print("status error:", resp.status)


async def reqs(url: str, method: str, count: int, limit: int) -> None:
    con = aiohttp.TCPConnector(limit=limit)
    async with aiohttp.ClientSession(connector=con, headers={
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:101.0) Gecko/20100101 Firefox/101.0",
        "Accept":     "*/*",
    }) as client:
        tasks = [asyncio.create_task(make_request(client, url, method))
                 for _ in range(count)]
        await asyncio.gather(*tasks)


def timer(func):
    def wrapper(*args, **kw):
        start = time.time()
        func(*args, **kw)
        end = time.time()
        print("time spent:\t%.2fs" % (end-start))
    return wrapper


@timer
def run(url: str, method: str = "GET", count: int = 100, limit: int = 64) -> None:
    """
    @description: 执行函数
    @param {str} url 
    @param {str} method 
    @param {int} count 
    @param {int} limit 最大并发数
    @return {*}
    """
    if os.name == 'nt':
        # windows 需要设置
        asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
    asyncio.run(reqs(url, method, count, limit))
    print("url:\t%s\nmethod:\t%s\ncount:\t%s\nlimit:\t%s" %
          (url, method, count, limit))


if __name__ == '__main__':

    # 请求本地构建的服务器
    # run(
    #     url="http://127.0.0.1:8080/",
    #     method="GET",
    #     count=10000,
    #     limit=64  # 最大并发数
    # )

    # 请求远程服务器
    run(
        url="https://www.jb51.net/",
        method="GET",
        count=500,
        limit=64  # 最大并发数
    )
